### ISSUE TITLE
an illegal memory access was encountered when using ClassNLLCriterion

### ISSUE THREAD

**Issue by Tae618** [torch/nn#1001]
When I use ClassNLLCriterion, there occurs a problem.
**THCudaCheck FAIL file=/home/yzw/torch/extra/cutorch/lib/THC/generic/THCStorage.c line=30 error=77 : an illegal memory access was encountered
/home/yzw/torch/install/bin/luajit: cuda runtime error (77) : an illegal memory access was encountered at /home/yzw/torch/extra/cutorch/lib/THC/generic/THCStorage.c:147**

I think it is not caused by indices in labels out of classes. The reason is 
When I run criterion:forward(logprobs, labels)  (note: I have used nn.MaskZeroCriterion, so zeroes in labels will be ignored.
logprobs size is  17x5x10893
labels is show in "print labels", there is no index out of 1~10893

[torch.CudaTensor of size 17x5x10893]

DEBUG> print torch.max(logprobs)
0
DEBUG> print torch.min(logprobs)
-9.8302659988403
DEBUG> print labels
  7111   7800   5973   2461   6915
  2768   3541   9351   3065   7943
  9351  10638   2461   8662     75
  2461  10893   2985   3418      0
     0      0   3418   8132      0
     0      0   1158      0      0
     0      0      0      0      0
     0      0      0      0      0
     0      0      0      0      0
     0      0      0      0      0
     0      0      0      0      0
     0      0      0      0      0
     0      0      0      0      0
     0      0      0      0      0
     0      0      0      0      0
     0      0      0      0      0
     0      0      0      0      0
[torch.CudaTensor of size 17x5] 

anyone have met such problem before?

**Comment by deepcoord** [comment]
I also got the same error when using nn.MaskZeroCriterion(nn.ClassNLLCriterion(),1).
Anyone have solved this problem?

**Comment by pavanky** [comment]
@deepcoord Sometimes the error message can be caught at a different place compared to where it is occuring because of the asynchronous nature of CUDA. Can you try putting `cutorch.synchronize()` after each module and see after which module exactly fails ?

**Comment by soumith** [comment]
another way to find out the exact failing module is:

```
CUDA_LAUNCH_BLOCKING=1 th yourscript.lua
```

But this is definitely a user-error. You are giving a label index to ClassNLLCriterion that is larger than the total number of classes, i.e. `outputs:size(2) < targets:max()`. Check that.

**Comment by pavanky** [comment]
> CUDA_LAUNCH_BLOCKING=1

I am not sure how I missed that in all these years.

**Comment by deepcoord** [comment]
@soumith I know there are some user-errors such as using label indexes larger than the total number of classes.
But this negative loss is not always occurred when running the same model with the same dataset.
If the reason is caused by larger label indexes, I must have had the error at a specific turn in an epoch.
I got it at the first try or the second, but I can run it over total epoches without the negative loss.
Here is my code snippet.

> local loss = criterion:forward(output, target)
> assert(loss>=0, "negative loss is occurred", print_status(loss, output, target))
> 
> function print_status(loss, output, target)
> print("loss = " .. loss)
> print(output)
> print(target)
> end

Here is the negative loss error, and its output and target.
The loss was -2.0618042044055e+16, and tensor size was 26x10x179 and 26x10, respectively.
FYI, the total number of classes are 179.
What am I missing?

> -- Epoch 1 / 100
> 
> loss = -2.0618042044055e+16
> (1,.,.) =
>  Columns 1 to 9
>  -5.2408 -5.2768 -5.1037 -5.1117 -5.2371 -5.1768 -5.1099 -5.0852 -5.1962
>  -5.2278 -5.2622 -5.1835 -5.0746 -5.2372 -5.2055 -5.1430 -5.0501 -5.2316
>  -5.2789 -5.2744 -5.1664 -5.1157 -5.1410 -5.1663 -5.1344 -5.0806 -5.2191
>  -5.1862 -5.2904 -5.1187 -5.2164 -5.1466 -5.1812 -5.1054 -5.1155 -5.1840
>  -5.2055 -5.2594 -5.0938 -5.1212 -5.2392 -5.2526 -5.1329 -5.1048 -5.1947
>  -5.2479 -5.2119 -5.1032 -5.1084 -5.1559 -5.2403 -5.0732 -5.0963 -5.2127
>  -5.1655 -5.2024 -5.0944 -5.1780 -5.1679 -5.1597 -5.1558 -5.1299 -5.1838
>  -5.1917 -5.3071 -5.0794 -5.1593 -5.1064 -5.1621 -5.1039 -5.1530 -5.2209
>  -5.2090 -5.2903 -5.0882 -5.0578 -5.1572 -5.1635 -5.1253 -5.1689 -5.1757
>  -5.1813 -5.2836 -5.1979 -5.0970 -5.1979 -5.1395 -5.1391 -4.9636 -5.2047
> 
> ...
> 
> Columns 172 to 179
>   0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
>   0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
>   0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
>   0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
>   0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
>   0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
>   0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
>  -5.2153 -5.1161 -4.9952 -5.2408 -5.2349 -5.2101 -5.1976 -5.0892
>   0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
>   0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
> [torch.CudaTensor of size 26x10x179]
> 
>   1   1   1   1   1   1   1   1   1   1
>  58   1   1  49   1   1   9   1   1   1
>   1   1  15  50   1   1  20   1   1   1
>   1   1   9   1   1  58   1  15  33   2
>   1   1  20   1   9  59  13   9   1  44
>   1   4  20   1   1  59  14  20   1  13
>   1  12   7   1   1   1  14   1  17   1
>   1   1   8  17   1   1  14   1   1   1
>   1   1   1   1  15   1   1   1  18   1
>   1  69   1  18  27   1   1  23  19  23
>   1   1   1   1   0  13   1  24   0  24
>   0   1   1   1   0  14   1   1   0   0
>   0   1   1   1   0  14   1   1   0   0
>   0  10  13  87   0  14   1   7   0   0
>   0  11  14   0   0  14  81   8   0   0
>   0   0  14   0   0  14  82   8   0   0
>   0   0  14   0   0  14   0   1   0   0
>   0   0  14   0   0  14   0   1   0   0
>   0   0  14   0   0   0   0   1   0   0
>   0   0   0   0   0   0   0  83   0   0
>   0   0   0   0   0   0   0  84   0   0
>   0   0   0   0   0   0   0   1   0   0
>   0   0   0   0   0   0   0  83   0   0
>   0   0   0   0   0   0   0  84   0   0
>   0   0   0   0   0   0   0  84   0   0
>   0   0   0   0   0   0   0  84   0   0
> [torch.CudaTensor of size 26x10]

**Comment by soumith** [comment]
@deepcoord the value 0 for indices is invalid. torch is one-indexed.

**Comment by deepcoord** [comment]
@soumith the value 0 for indices are just padding since I am using MaskZero in Sequencer. So I don't think I have used wrong indexes
